Add Python 3 support

chiwanpark opened the pull request:
Python 3 지원을 위해 아래 사항을 수정했습니다.
* `six` 라이브러리 추가 (py2/3 호환용, MIT license)
* 코드가 py3에서도 동작하도록 수정
* `y_vocab.cPickle` 파일 다시 만들어야하는 부분 README에 추가

기타 수정 사항
* `config.json`의 오타 수정 (`preidct` -> `predict`)
* Python 3.5에서만 발생하는 버그로 인해 `map_async` 결과 timeout 시간 감소

chiwanpark committed:
Add Python 3 compatibility

  - Add six to requirements.txt and NOTICE.md
  - Fix typo in some codes

chiwanpark committed:
Decrease timeout for map_async due to a weird bug in Python 3.5

  - See https://bugs.python.org/issue25155

ummae commented:
@chiwanpark 👍 
가능하다면 `y_vocab.cPickle` 생성한 파일도 이름이 겹치지 않도록 해서 같이 포함하면 좋을 것 같아요.

GzuPark commented on config.json:
```diff
@@ -5,7 +5,7 @@
     "db_chunk_size": 100000,
     "num_workers": 10,
-    "num_preidct_workers": 2,
+    "num_predict_workers": 2,
```
감사합니다 👍

chiwanpark committed:
Use y_vocab.py3.cPickle in Python 3

chiwanpark committed:
Add y_vocab.py3.cPickle

chiwanpark commented:
@ummae Python 3에서 사용할 `y_vocab.py3.cPickle` 파일 추가했습니다. 코드에서도 자동으로 Python 3인 경우 해당 파일을 불러오도록 수정했어요.

GzuPark commented:
@chiwanpark 👏 
README.md의 실행방법 4. 에서도 안내를 해주는게 좋지 않을까요?
예를 들면,
```
python2 evaluate.py evaluate predict.tsv ./data/train/data.h5py dev ./data/y_vocab.cPickle
python3 evaluate.py evaluate predict.tsv ./data/train/data.h5py dev ./data/y_vocab.py3.cPickle
```

chiwanpark committed:
precess -> process

chiwanpark committed:
Use mmh3 instead of python builtin hash to avoid randomized seed

ummae commented on data.py:
```diff
@@ -25,7 +24,10 @@
 import fire
 import h5py
 import numpy as np
+import mmh3
```
버저닝을 하는 게 좋을 것 같습니다. python2로 사용중이던 사람들은 해시 값이 달라져서 코드 업데이트 후에는 기존 데이터를 사용못하게 되네요.

대략 이런 느낌으로
```
hash = lambda x: mmh3.hash(x, seed=17) if six.PY3 else hash

chiwanpark replied:
@ummae 해당 부분 수정했습니다.

ummae reviewed (changes requested):
@chiwanpark

ummae committed:
predict chunk by chunk

chiwanpark committed:
Choice hash function depending on Python version

ummae committed:
fix a bug, update progress bar with more info

chiwanpark committed:
Make data generator thread-safe

  - Keras requires a thread-safe generator
  - See https://github.com/keras-team/keras/issues/1638

chiwanpark committed:
Update README.md

chiwanpark commented:
위 코드로 학습시킨 모델로 dev 계산하여 채점하면 아래와 같은 점수를 얻을 수 있습니다.

* Python 2: 0.851424
* Python 3: 0.845507

버전에 따른 점수 차이는 Python 버전 별 해시 함수 차이(built-in hash vs mmh3)로 생각됩니다.

ummae commented:
👍
